File: ambeth/Ambeth.Merge/ambeth/bytecode/EntityEnhancementHint.cs
using System;

namespace De.Osthus.Ambeth.Bytecode
{
    public sealed class EntityEnhancementHint : IEnhancementHint
    {
        public static readonly EntityEnhancementHint HOOK = new EntityEnhancementHint();

        private EntityEnhancementHint()
        {
            // Intended blank
        }

        public T Unwrap<T>() where T : IEnhancementHint
        {
            return (T)Unwrap(typeof(T));
        }

        public Object Unwrap(Type includedContextType)
        {
            if (typeof(EntityEnhancementHint).IsAssignableFrom(includedContextType))
            {
                return this;
            }
            return null;
        }
    }
}

File: jambeth/jAmbeth.Merge/src/de/osthus/ambeth/bytecode/EntityEnhancementHint.java
package de.osthus.ambeth.bytecode;

public enum EntityEnhancementHint implements IEnhancementHint
{
	HOOK;

	@SuppressWarnings("unchecked")
	@Override
	public <T extends IEnhancementHint> T unwrap(Class<T> includedContextType)
	{
		if (EntityEnhancementHint.class.isAssignableFrom(includedContextType))
		{
			return (T) this;
		}
		return null;
	}

	@Override
	public String toString()
	{
		return getClass().getName();
	}
}
